src/Core/Block.cpp

```diff
@@ -145,15 +145,17 @@ void Block::insert(size_t position, ColumnWithTypeAndName elem)
     if (elem.name.empty())
         throw Exception("Column name in Block cannot be empty", ErrorCodes::AMBIGUOUS_COLUMN_NAME);
 
-    for (auto & name_pos : index_by_name)
-        if (name_pos.second >= position)
-            ++name_pos.second;
-
     auto [it, inserted] = index_by_name.emplace(elem.name, position);
     if (!inserted)
         checkColumnStructure<void>(data[it->second], elem,
             "(columns with identical name must have identical structure)", true, ErrorCodes::AMBIGUOUS_COLUMN_NAME);
 
+    for (auto itr = index_by_name.begin(); itr != index_by_name.end(); ++itr)
+    {
+        if (itr->second >= position && (!inserted || itr != it))
```

Comment:
But we already checked for `!inserted` above.

Reply:
Because this is the common logic for the cases:
1) inserted success
2) not inserted, but checkColumnStructure() was ok and didn't throw exception

Reply:
Ok.